Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,459 @@
+        case .startLoop:
+            break
+        }
+
+        while true {
+            let action = self.nextPollLoopAction()
+
+            switch action {
+            case .pollAndSleep(let pollClosure):
+                // Poll Kafka for new acknowledgements and sleep for the given
+                // pollInterval to avoid hot looping.
+                pollClosure?()
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    self.terminate()
+                    throw error
+                }
+            case .suspendPollLoop:
+                // The downstream consumer asked us to stop sending new messages.
+                // We therefore await until we are unsuspended again.
+                try await withTaskCancellationHandler {
+                    // Necessary.
+                    // We don't want to create a leaking continuation
+                    // when we are already cancelled.
+                    // The cancellation of a Task does not mean
+                    // its code will not still be executed.
+                    guard Task.isCancelled == false else {
+                        self.terminate(CancellationError())
+                        throw CancellationError()
+                    }
+                    try await withCheckedThrowingContinuation { continuation in
+                        self.stateMachine.withLockedValue { $0.suspendLoop(continuation: continuation) }
+                    }
+                } onCancel: {
+                    self.terminate(CancellationError())
+                    return
+                }
+            case .shutdownPollLoop:
+                // We have been asked to close down the poll loop.
+                self.terminate()
+                return
+            }
+        }
+    }
+
+    /// Yield new elements to the underlying `NIOAsyncSequenceProducer`.
+    func yield(_ element: Element) {
+        let action = self.stateMachine.withLockedValue { $0.yield(element) }
+        switch action {
+        case .produceMore:
+            self.produceMore()
+        case .stopProducing:
+            self.stopProducing()
+        case .terminate:
+            self.terminate()
+        case .none:
+            break
+        }
+    }
+
+    /// Determines the next action to be taken in the poll loop based on the current state.
+    ///
+    /// - Returns: The next action for the poll loop.
+    func nextPollLoopAction() -> KafkaPollingSystem.StateMachine.PollLoopAction {
+        return self.stateMachine.withLockedValue { $0.nextPollLoopAction() }
+    }
+
+    /// Stop producing new elements to the
+    /// `source` ``NIOAsyncSequenceProducer``.
+    private func stopProducing() {
+        self.stateMachine.withLockedValue { $0.stopProducing() }
+    }
+
+    /// Kill the ``KafkaPollingSystem``'s poll loop and free its resources.
+    func terminate(_ error: Error? = nil) {
```

Comment:
Can this be private?

Reply:
Yes this would be no problem because we can also invoke `didTerminate()` from outside, but I thought that the name `terminate()` made more sense, also because `didTerminate()` is a delegate method.

(My implementation needs an `internal` method for terminating the `KafkaPollingSystem` because I invoke this from `KafkaPollingSystem.shutDownGracefully`)